samples/StaticFilesAuth/Startup.cs

```diff
@@ -114,81 +138,61 @@ public void Configure(IApplicationBuilder app, IHostingEnvironment env, IAuthori
             });
         }
 
-        // Blanket authorization, any authenticated user is allowed access to these resources.
-        private static void MapAuthenticatedFiles(IApplicationBuilder branch, PhysicalFileProvider files)
+        private void SetupFileServer(IApplicationBuilder builder, IFileProvider files)
         {
-            branch.Use(async (context, next) =>
-            {
-                if (!context.User.Identity.IsAuthenticated)
-                {
-                    await context.ChallengeAsync(new AuthenticationProperties()
-                    {
-                        // https://github.com/aspnet/Security/issues/1730
-                        // Return here after authenticating
-                        RedirectUri = context.Request.PathBase + context.Request.Path + context.Request.QueryString
-                    });
-                    return;
-                }
-
-                await next();
-            });
-            branch.UseFileServer(new FileServerOptions()
+            builder.UseFileServer(new FileServerOptions()
             {
                 EnableDirectoryBrowsing = true,
                 FileProvider = files
             });
         }
 
-        // Policy based authorization, requests must meet the policy criteria to be get access to the resources.
-        private static void MapImperativeFiles(IAuthorizationService authorizationService, IApplicationBuilder branch, PhysicalFileProvider files)
+        private static void SetFileEndpoint(HttpContext context, PhysicalFileProvider files, PathString filePath, string policy)
         {
-            branch.Use(async (context, next) =>
+            var fileSystemInfo = GetFileSystemInfo(files, filePath);
+            if (fileSystemInfo != null)
             {
-                var fileInfo = files.GetFileInfo(context.Request.Path);
-                AuthorizationResult result = null;
-                if (fileInfo.Exists)
-                {
-                    result = await authorizationService.AuthorizeAsync(context.User, fileInfo, "files");
-                }
-                else
-                {
-                    // https://github.com/aspnet/Home/issues/2537
-                    var dir = files.GetDirectoryContents(context.Request.Path);
-                    if (dir.Exists)
-                    {
-                        result = await authorizationService.AuthorizeAsync(context.User, dir, "files");
-                    }
-                    else
-                    {
-                        context.Response.StatusCode = StatusCodes.Status404NotFound;
-                        return;
-                    }
-                }
+                var metadata = new List<object>();
+                metadata.Add(fileSystemInfo);
+                metadata.Add(new AuthorizeAttribute(policy));
+
+                var endpoint = new Endpoint(
+                    c => Task.CompletedTask,
```

Comment:
null? Otherwise this breaks 404.
It's a bit awkward to use Endpoints for auth without using them to actually handle the request.

Reply:
It doesn't break 404 because an endpoint is only being created for a file, and files are always being return by UseFileSystem middleware.

The endpoint request delegate is never being called, but if it was then returning a null would NRE when it is awaited.

Reply:
It should never be awaited. If it matches a file then this would never get run. If it doesn't match a file then calling this would be wrong.
https://github.com/aspnet/HttpAbstractions/pull/1059/files#diff-3376870948b08e8c913217b7dab1bdcbR86

Reply:
Better yet, since we don't expect it to get called it should throw InvalidOperationException.